### ISSUE TITLE
`--since` flag to only flag files not modified in git in certain period of time

### ISSUE THREAD

**Issue by me4502** [webpro/knip#204]
Sometimes code is added during a project that's not currently used, but will be used in a follow up PR. When cleaning up the codebase, I generally check with git-blame in an IDE that the file or method haven't been modified recently before removing them, to not remove code that someone is working on.

My proposal is to add a `--since` flag, used like `--since=2w` to not flag any files that have been modified in git within the past 2 weeks for example. 

A possible extension is to expand this to checking per-line history for exports/functions/etc, however I feel this might have too significant an impact on performance. I feel this would provide significant benefit by just checking file modified time from a benefit to time perspective, especially for checks like unused files.

**Comment by webpro** [comment]
Not sure whether this workflow is something Knip should get into/promote.

On a file level, might make sense to use `ignore` in a `knip.jsonc` or `knip.ts`, add comments, and manage those things in a central place.

You could also implement a custom reporter (which can receive extra arguments). Eventually that could end up as a new internal or external reporter for others to use as well.

**Comment by me4502** [comment]
Definitely understand it might not make sense as a direct feature. The ignore option within `knip.ts` I could see working, would just require basically scanning the git log and generating a big ignore list beforehand. Not super ideal as it does feel like it'd slow down both startup and the running of Knip (as I've noticed that large numbers of ignores do seem to slow down runtime rather substantially)

In terms of a custom reporter, my only concern would be that it feels more like a report preprocessor than an actual outputter. Can custom reporters work in a chain? As in, could this reporter "filter" the data, then pass it on to another reporter?

**Comment by webpro** [comment]
At this point all it does is call the default exported function with a bunch of data and some extra options and flags: https://github.com/webpro/knip/blob/93fecdabcdcc2324c69256bce3d886cbc3fb8046/src/cli.ts#L63-L72

No chaining or anything, just a single function call.

Adding `--preprocessor` or similar is trivial, planning to add that.

Not aware of anyone using custom reporter(s), but adding this feature will emphasize that changes in this data structure are breaking changes.

As a side note, should also support multiple reporters (e.g. one to write to files and one to print stuff).

**Comment by webpro** [comment]
I've released [v2.20.0-preprocess.0](https://github.com/webpro/knip/releases/tag/2.20.0-preprocess.0). You can install using e.g. `npm install -D knip@preprocess`

This is a working example usage from the Knip repo itself:

```
$ knip \
  --reporter ./fixtures/reporters/minimal.ts \
  --preprocessor ./fixtures/preprocessors/identity.ts \
  --preprocessor ./fixtures/preprocessors/minimal.ts \
  --reporter compact \
  --reporter json
files 1
dependencies 0
devDependencies 0
unlisted 0
binaries 0
unresolved 0
exports 0
nsExports 0
types 0
nsTypes 0
enumMembers 0
classMembers 0
duplicates 0
Unused files (1)
file:///Users/lars/p/knip/knip/fixtures/preprocessors/minimal.ts
[{"file":"file:/Users/lars/p/knip/knip/fixtures/preprocessors/minimal.ts","files":true,"dependencies":[],"devDependencies":[],"unlisted":[],"unresolved":[],"exports":[],"types":[],"enumMembers":{},"classMembers":{},"duplicates":[]}]

```

Would love to gather some feedback about API and usage before finalizing with docs etc. and merge to main. WDYT?

**Comment by me4502** [comment]
Thanks for doing that, and sorry for the delay in my response. That preprocess API looks mostly perfect for this, although I feel given file IO could be fairly common, having it async might make a lot of sense?

It'd be usable while not async, but I feel it could be cleaner to write with and potentially much faster if async. Thanks either way though :)

**Comment by webpro** [comment]
Both preprocessors and reporters can be async :)

(the functions are `await`-ed here: https://github.com/webpro/knip/commit/444dfc668c50b5dfe9543cd05a1eb37d942b1904#diff-fa8d4e24d8399e8350f1c8bad05df53e8032ea995835bf911507015e2db61cddR85-R87)

**Comment by me4502** [comment]
Thanks, sorry I had just checked the type definition and assumed it couldn't be a promise based on that. The preprocess API seems perfect for this :)

**Comment by webpro** [comment]
:rocket: _This issue has been resolved in v2.21.0. See [Release 2.21.0](https://github.com/webpro/knip/releases/tag/2.21.0) for release notes._

**Comment by webpro** [comment]
No worries, it's out now 🐣
